@@ -, +, @@ 
---
 drivers/net/wireless/mwifiex/usb.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)
--- a/drivers/net/wireless/mwifiex/usb.c	
+++ a/drivers/net/wireless/mwifiex/usb.c	
@@ -235,9 +235,11 @@ setup_for_next:
 	if (card->rx_cmd_ep == context->ep) {
 		mwifiex_usb_submit_rx_urb(context, size);
 	} else {
-		context->skb = NULL;
-		if (atomic_read(&adapter->rx_pending) <= HIGH_RX_PENDING)
+		if (atomic_read(&adapter->rx_pending) <= HIGH_RX_PENDING) {
 			mwifiex_usb_submit_rx_urb(context, size);
+		} else {
+			context->skb = NULL;
+		}
 	}
 
 	return;
-- 