diff -Naur linux-4.4-rc7.old/debian.master/config/config.common.ubuntu linux-4.4-rc7/debian.master/config/config.common.ubuntu
--- linux-4.4-rc7.old/debian.master/config/config.common.ubuntu	2016-01-03 17:57:13.827752736 +0000
+++ linux-4.4-rc7/debian.master/config/config.common.ubuntu	2016-01-03 20:55:21.802084654 +0000
@@ -695,6 +695,24 @@
 CONFIG_AUDIT_GENERIC=y
 CONFIG_AUDIT_TREE=y
 CONFIG_AUDIT_WATCH=y
+CONFIG_AUFS_BDEV_LOOP=y
+# CONFIG_AUFS_BRANCH_MAX_1023 is not set
+CONFIG_AUFS_BRANCH_MAX_127=y
+# CONFIG_AUFS_BRANCH_MAX_32767 is not set
+# CONFIG_AUFS_BRANCH_MAX_511 is not set
+# CONFIG_AUFS_BR_FUSE is not set
+CONFIG_AUFS_BR_HFSPLUS=y
+# CONFIG_AUFS_BR_RAMFS is not set
+# CONFIG_AUFS_DEBUG is not set
+CONFIG_AUFS_EXPORT=y
+# CONFIG_AUFS_FHSM is not set
+CONFIG_AUFS_FS=m
+# CONFIG_AUFS_HNOTIFY is not set
+CONFIG_AUFS_INO_T_64=y
+# CONFIG_AUFS_RDU is not set
+CONFIG_AUFS_SBILIST=y
+# CONFIG_AUFS_SHWH is not set
+# CONFIG_AUFS_XATTR is not set
 CONFIG_AURORA_NB8800=m
 CONFIG_AUTO_ZRELADDR=y
 CONFIG_AX25=m
